### PR TITLE
Add DHCP Server to companion

### LINKED ISSUE #287: No DHCP Server on Cable Guy

### PR DESCRIPTION
**Pull request by rafaellehmkuhl**
Fix #287 

- Add dnsmasq as DHCP Server provider
- Use 192.168.2.2 as gateway
- Offer IPs in the 192.168.2.100-200 range
- Changes default  GCS Link to 192.168.2.2
- Add proper logging to cable-guy

### CODE REVIEW

**Review by patrickelectric — CHANGES_REQUESTED** [review]

File: install/network/dnsmasq.sh
Comment (patrickelectric) [reply install/network/dnsmasq.sh:13]: Can it run in the dockerfile ?

File: install/network/dnsmasq.sh
Comment (rafaellehmkuhl) [reply install/network/dnsmasq.sh:22]: Oh, yeah. This really passed by. I had a script before, but them I decided to replicate the one we are using with Avahi for maintaining purposes and totally didn't realized that. I will check for the config file.

File: install/network/dnsmasq.sh
Comment (rafaellehmkuhl) [reply install/network/dnsmasq.sh:13]: For the first part, you mean installing from somewhere (maybe even the debian repository) but avoiding the apt tool?

Comment (patrickelectric) [reply install/network/dnsmasq.sh:13]: we can't use debian repository, with that you'll need to install to install others tools to install .deb files under non-debian  distros.

I believe that the best solution, if possible, is to install it under docker.

Comment (rafaellehmkuhl) [reply install/network/dnsmasq.sh:13]: I really don't think it makes much sense to run a host os service under a docker container...

Besides the logical view, that would also add a good amount of work, for a problem that its not a priori, and that we currently have already (g++, python3-dev, python3-pip and python3-setuptools are all installed with apt).

I believe we should merge this and think about a final solution for all packages together, not for this one specifically.

Comment (patrickelectric) [reply install/network/dnsmasq.sh:13]: > apt

Inside docker, this script runs in the host computer, and you can't guarantee that apt will be available.

Comment (rafaellehmkuhl) [reply install/network/dnsmasq.sh:13]: Yes, but as with them, we can't guarantee that apt will be available either.

What does the fact that they run inside a container changes?

Comment (patrickelectric) [reply install/network/dnsmasq.sh:13]: > we can't guarantee that apt will be available either.

Yes we can, we use debian based docker images. 

The host computer can run debian, arch or any other distro that the user desires. We have control over the docker, so we can guarantee that ap exists.

Comment (Williangalvani) [reply install/network/dnsmasq.sh:13]: > Besides the logical view, that would also add a good amount of work, for a problem that its not a priori, and that we currently have already (g++, python3-dev, python3-pip and python3-setuptools are all installed with apt).

Those run inside the debian container. We know the container is debian, but we can't say the host os is, too.

Comment (rafaellehmkuhl) [reply install/network/dnsmasq.sh:13]: Oh, you're problem is with the os itself, not the apt tool. Got it.

**Comment by rafaellehmkuhl** [comment]
Did some changes on the PR today so it becomes more consistent with the current architecture and allow a future hotspot PR to  be enabled without much hassle.

I consider it ready and tested. Image available under rafaellehmkuhl/companion-core:dhcp.

File: core/tools/dnsmasq/bootstrap.sh
Comment (Williangalvani) [inline core/tools/dnsmasq/bootstrap.sh:10]: this looks weird. is the install step starting the service? even then, this runs in build time, I expected something analog to `service disable", not stop.

can you make explain better in comments?

File: core/services/ardupilot_manager/ArduPilotManager.py
Comment (Williangalvani) [inline core/services/ardupilot_manager/ArduPilotManager.py:116]: this will break all current setups out there, right? we should talk about about it. Ideally we should be able to make both work. If we can't, this is the priority.

**Review by Williangalvani — CHANGES_REQUESTED** [review]
this is looking pretty good. I got a few comments though.

Can you upload an image for testing?

**Comment by rafaellehmkuhl** [comment]
> this is looking pretty good. I got a few comments though.
> 
> Can you upload an image for testing?

Image online on rafaellehmkuhl/companion-core:dhcp

File: core/services/ardupilot_manager/ArduPilotManager.py
Comment (rafaellehmkuhl) [reply core/services/ardupilot_manager/ArduPilotManager.py:116]: Yes, altough current setups are all for testing :smile: 

I'm inclined to say it's a decision better to took now them later.

File: bootstrap/bootstrap/bootstrap.py
Comment (rafaellehmkuhl) [reply bootstrap/bootstrap/bootstrap.py:146]: Fair. Will check.

Can we maintain this but remove the port bind?

File: core/tools/dnsmasq/bootstrap.sh
Comment (rafaellehmkuhl) [reply core/tools/dnsmasq/bootstrap.sh:10]: I'm not much familiar with non-systemd managers, as this one, and it doesn't have 'disable'.

During tests I remember to find out that maybe the stop command also disabled the service, but I can be wrong. Will test without it.

File: core/services/cable_guy/main.py
Comment (patrickelectric) [inline core/services/cable_guy/main.py:31]: `from logging import LogRecord`

**Review by patrickelectric — CHANGES_REQUESTED** [review]

File: core/services/cable_guy/main.py
Comment (patrickelectric) [reply core/services/cable_guy/main.py:31]: Also, check: loguru_mypy

Comment (rafaellehmkuhl) [reply core/services/cable_guy/main.py:31]: This is basically the same discussion we had in the SITL PR. When introduced, LogRecord raises a lot of other typing problems, that I'm discussing with the Loguru author.

As its replicated code with ArduPilotManager, can we keep it like it is and solve both together on the common library?

File: core/services/cable_guy/main.py
Comment (rafaellehmkuhl) [reply core/services/cable_guy/main.py:25]: The opposite, but like on SITL I will remove this option.

File: core/services/cable_guy/main.py
Comment (patrickelectric) [reply core/services/cable_guy/main.py:31]: I got it working here without problems only doing what I said

Comment (rafaellehmkuhl) [reply core/services/cable_guy/main.py:31]:
```
core/services/cable_guy/main.py:27: error: Incompatible types in assignment (expression has type "int", variable has type "str")
Found 1 error in 1 file (checked 1 source file)
```

You're sure you're not receiving this as soon as LogRecord is introduced?

Comment (patrickelectric) [reply core/services/cable_guy/main.py:31]: I just used: level: Union[int, str]

Comment (rafaellehmkuhl) [reply core/services/cable_guy/main.py:31]: Tracked on #313

**Comment by rafaellehmkuhl** [comment]
Fix #313

File: core/tools/dnsmasq/bootstrap.sh
Comment (rafaellehmkuhl) [reply core/tools/dnsmasq/bootstrap.sh]: Testes again and its not needed indeed. Removed.

File: bootstrap/bootstrap/bootstrap.py
Comment (rafaellehmkuhl) [reply bootstrap/bootstrap/bootstrap.py:146]: Maintained function but removed bind. Tested and working.

File: core/services/ardupilot_manager/ArduPilotManager.py
Comment (jaxxzer) [reply core/services/ardupilot_manager/ArduPilotManager.py:116]: let's keep the dhcp server running at 192.168.2.2, like companion 0

**Review by patrickelectric — CHANGES_REQUESTED** [review]

File: bootstrap/test_bootstrap.py
Comment (rafaellehmkuhl) [reply bootstrap/test_bootstrap.py:33]: The consensus was to maintain it, but I can remove, no problem.

File: core/tools/dnsmasq/bootstrap.sh
Comment (rafaellehmkuhl) [reply core/tools/dnsmasq/bootstrap.sh:16]: It's checks if the installation process indeed created the original file, preventing changes in this behavior to silent affect out application.

If you find it unnecessary, there's no problem in removing it.

Comment (patrickelectric) [reply core/tools/dnsmasq/bootstrap.sh:16]: Can you just add a comment about it ? The logic behind the code is not clear without this explanation.

File: bootstrap/test_bootstrap.py
Comment (Williangalvani) [reply bootstrap/test_bootstrap.py:33]: we talked about it and I agree this should be introduced.
It IS out of scope for this pr, but it should get in eventually.
I have no strong feelings about keeping it here or not.

File: core/services/cable_guy/api/manager.py
Comment (rafaellehmkuhl) [reply core/services/cable_guy/api/manager.py:104]: The DHCP Server is always on, yes.

File: core/services/cable_guy/api/manager.py
Comment (rafaellehmkuhl) [reply core/services/cable_guy/api/manager.py:308]: Yes, that's the correct behavior.

File: core/services/ardupilot_manager/main.py
Comment (patrickelectric) [reply core/services/ardupilot_manager/main.py:38]: Yes, where the code that you are fixing was provided.
As general rule, fixing commit that was not merged yet should be done in the same commit and not in a new one. 
Otherwise you are creating unnecessary commits.

Comment (rafaellehmkuhl) [reply core/services/ardupilot_manager/main.py:38]: Your comment is on the ardupilot-manager service, not on cable-guy.

Comment (rafaellehmkuhl) [reply core/services/ardupilot_manager/main.py:38]: They are two different code bases.

File: core/tools/dnsmasq/dnsmasq.conf
Comment (patrickelectric) [inline core/tools/dnsmasq/dnsmasq.conf:11]: Mixed feelings about enabling ipv6, in the end it only results in problems, I think that would be more sane to just disable it.

Comment (rafaellehmkuhl) [reply core/tools/dnsmasq/dnsmasq.conf:11]: Can you test without it?

I'm not a DHCP expert, so I don't know if this is a regular behavior, but for my machine, not having a IPv6 provided when the machine asked for it would prevent also the IPv4 to be provided.

Comment (patrickelectric) [reply core/tools/dnsmasq/dnsmasq.conf:11]: I just removed and appears to be working.

Comment (rafaellehmkuhl) [reply core/tools/dnsmasq/dnsmasq.conf:11]: Will check if there's anything on my machine regarding this configuration and test with another one.

File: core/tools/dnsmasq/dnsmasq.conf
Comment (Williangalvani) [inline core/tools/dnsmasq/dnsmasq.conf:13]: ask?

File: core/tools/dnsmasq/dnsmasq.conf
Comment (rafaellehmkuhl) [reply core/tools/dnsmasq/dnsmasq.conf:11]: On tests, although I can see the message on failure to provide an IPv6, this time the IPv4 got leased, so I assume it was a coincidence what I have seem previously. 
Removing.

File: core/tools/dnsmasq/dnsmasq.conf
Comment (rafaellehmkuhl) [reply core/tools/dnsmasq/dnsmasq.conf]: Fixed.

**Comment by Williangalvani** [comment]
I don't get it. IS the DHCP server always on? The user should be able to disable it.
I expect a few devices will behave differently in each case.

**Comment by rafaellehmkuhl** [comment]
> I don't get it. IS the DHCP server always on? The user should be able to disable it.
> I expect a few devices will behave differently in each case.

Right now it's always on, yes.
Do you some specific problem with this behavior?

I was thinking about it and didn't see an use case on a first moment.

My concern on a first moment, that made me not enable/disable dnsmasq was that I would not want cable-guy for disabling the wifi DHCP server when we implemented it, but sometime after I realized we can have two DHCP server (dnsmasq instances) running separated, one for ethernet and another for wifi.

File: core/services/cable_guy/api/manager.py
Comment (patrickelectric) [inline core/services/cable_guy/api/manager.py:140]: Try to avoid such changes in a non related commit.

File: core/tools/dnsmasq/bootstrap.sh
Comment (patrickelectric) [inline core/tools/dnsmasq/bootstrap.sh:16]: Since the important thing here is the binary itself, we may need just to check the binary installation or if the binary itself exists.

File: core/tools/dnsmasq/bootstrap.sh
Comment (patrickelectric) [inline core/tools/dnsmasq/bootstrap.sh:21]: This is not necessary since we can provide the path for the configuration file with `--conf-file` as an argument.

File: core/tools/dnsmasq/bootstrap.sh
Comment (patrickelectric) [inline core/tools/dnsmasq/bootstrap.sh:24]: Why is this necessary ? Will just generate unnecessary noise.

**Review by patrickelectric — CHANGES_REQUESTED** [review]
I'm not sure if I'm ok with it, having the DHCP server always on could result in problems under the network or in unreliable behavior when using unmanaged mode.

**Comment by rafaellehmkuhl** [comment]
> I'm not sure if I'm ok with it, having the DHCP server always on could result in problems under the network or in unreliable behavior when using unmanaged mode.

Yes. I've discussed that with @Williangalvani and we decided to change the behavior. Forgot to comment here, my bad.

I'm just waiting for the approval of #330 right now, since the manager for `dnsmasq` will be there.

**Comment by patrickelectric** [comment]
>I'm just waiting for the approval of #330 right now, since the manager for dnsmasq will be there.

what ? Why dnsmasq should be in commonwealth ?

**Comment by rafaellehmkuhl** [comment]
> > I'm just waiting for the approval of #330 right now, since the manager for dnsmasq will be there.
> 
> what ? Why dnsmasq should be in commonwealth ?

Dnsmasq will also be used on a separate instance for the Wifi DHCP server.

**Comment by rafaellehmkuhl** [comment]
> > > I'm just waiting for the approval of #330 right now, since the manager for dnsmasq will be there.
> > 
> > 
> > what ? Why dnsmasq should be in commonwealth ?
> 
> Dnsmasq will also be used on a separate instance for the Wifi DHCP server.

To better explain it:

We should have a wifi hotspot provided by the wifi service soon. This hotspot will also need a DHCP server. So we can keep both services completely separated, with one not interfeiring on the other, the idea is for each of them to control their own instance of dnsmasq (DHCP server provider), with its own configuration.

**Comment by rafaellehmkuhl** [comment]
Finished and tested.

The image is available for tests under `rafaellehmkuhl/companion-core:dhcp`.

File: core/services/cable_guy/api/manager.py
Comment (patrickelectric) [inline core/services/cable_guy/api/manager.py:55]: I think that we should allow any IP address but have this one as default for bluerov and frontend.

File: core/libs/commonwealth/commonwealth/utils/DHCPServerManager.py
Comment (rafaellehmkuhl) [reply core/libs/commonwealth/commonwealth/utils/DHCPServerManager.py:32]: Agree. Will just test for the error-code.

File: core/libs/commonwealth/commonwealth/utils/DHCPServerManager.py
Comment (rafaellehmkuhl) [reply core/libs/commonwealth/commonwealth/utils/DHCPServerManager.py:31]: Yes.

File: core/services/cable_guy/api/manager.py
Comment (rafaellehmkuhl) [reply core/services/cable_guy/api/manager.py:55]: I think that would be cool, but this will add some good work and test.

Can we open an enhancement issue to track that and postpone it? I would like to merge the DHCP server as it is, as it adds value already, and focus on the firmware management and frontend for now.

Comment (rafaellehmkuhl) [reply core/services/cable_guy/api/manager.py]: Opened #336 to track custom configurations.

**Review by patrickelectric — APPROVED** [review]

**Review by Williangalvani — APPROVED** [review]

**Pull request merged**
